File: src/scorer/election-assessors/lordship.ts
import { RASI_NAMES } from '../../data/rasis';
import type { ElectionChartSnapshot } from '../../lib/election-chart-api';
import type { PrimitiveOutcome } from './contracts';
import { completePlanetPositions } from './event-admission';

export const LORD_POSITION_FACTS_UNAVAILABLE =
  'Complete canonical nine-graha position facts are unavailable or conflicting.';

const CANONICAL_GRAHAS = new Set([
  'Surya', 'Chandra', 'Kuja', 'Budha', 'Guru', 'Shukra', 'Shani', 'Rahu', 'Ketu',
]);

export const CLASSICAL_RASI_LORDS = Object.freeze({
  Mesha: 'Kuja',
  Vrishabha: 'Shukra',
  Mithuna: 'Budha',
  Karka: 'Chandra',
  Simha: 'Surya',
  Kanya: 'Budha',
  Tula: 'Shukra',
  Vrischika: 'Kuja',
  Dhanu: 'Guru',
  Makara: 'Shani',
  Kumbha: 'Shani',
  Meena: 'Guru',
} as const satisfies Readonly<Record<string, string>>);

export function deriveLagnaSixthLords(lagnaRashi: string): {
  lagnaLord: string;
  sixthRashi: string;
  sixthLord: string;
} | null {
  const lagnaIndex = RASI_NAMES.indexOf(lagnaRashi);
  if (lagnaIndex < 0) return null;
  const sixthRashi = RASI_NAMES[(lagnaIndex + 5) % 12];
  return {
    lagnaLord: CLASSICAL_RASI_LORDS[
      lagnaRashi as keyof typeof CLASSICAL_RASI_LORDS
    ],
    sixthRashi,
    sixthLord: CLASSICAL_RASI_LORDS[
      sixthRashi as keyof typeof CLASSICAL_RASI_LORDS
    ],
  };
}

export function wholeSignShortestDistance(
  leftRashi: string,
  rightRashi: string,
): number | null {
  const left = RASI_NAMES.indexOf(leftRashi);
  const right = RASI_NAMES.indexOf(rightRashi);
  if (left < 0 || right < 0) return null;
  const delta = Math.abs(left - right);
  return Math.min(delta, 12 - delta);
}

export function evaluateRasiLordSeparation(
  chart: ElectionChartSnapshot,
  options: {
    firstRashi: string;
    secondRashi: string;
    requiredShortestDistance: number;
  },
): PrimitiveOutcome {
  if (
    !options || typeof options !== 'object' || Array.isArray(options)
    || !RASI_NAMES.includes(options.firstRashi)
    || !RASI_NAMES.includes(options.secondRashi)
    || !Number.isInteger(options.requiredShortestDistance)
    || options.requiredShortestDistance < 0
    || options.requiredShortestDistance > 6
  ) {
    return {
      status: 'unknown',
      evidence: ['The Rasi-lord separation configuration is incomplete.'],
    };
  }
  if (
    !chart || typeof chart !== 'object' || Array.isArray(chart)
    || !Array.isArray(chart.planets)
    || Array.from({ length: chart.planets.length }, (_, index) => chart.planets[index])
      .some(planet => !planet)
  ) {
    return { status: 'unknown', evidence: [LORD_POSITION_FACTS_UNAVAILABLE] };
  }
  const positions = completePlanetPositions(chart, CANONICAL_GRAHAS);
  if (!positions) {
    return { status: 'unknown', evidence: [LORD_POSITION_FACTS_UNAVAILABLE] };
  }

  const firstLord = CLASSICAL_RASI_LORDS[
    options.firstRashi as keyof typeof CLASSICAL_RASI_LORDS
  ];
  const secondLord = CLASSICAL_RASI_LORDS[
    options.secondRashi as keyof typeof CLASSICAL_RASI_LORDS
  ];
  const firstPosition = positions.get(firstLord)!;
  const secondPosition = positions.get(secondLord)!;
  const distance = wholeSignShortestDistance(firstPosition.rashi, secondPosition.rashi);
  if (distance === null) {
    return { status: 'unknown', evidence: [LORD_POSITION_FACTS_UNAVAILABLE] };
  }

  const evidence = firstLord === secondLord
    ? `${options.firstRashi} and ${options.secondRashi} share lord ${firstLord} in ${firstPosition.rashi}; shortest Whole Sign distance is 0 (required ${options.requiredShortestDistance}).`
    : `${options.firstRashi} lord ${firstLord} is in ${firstPosition.rashi}; ${options.secondRashi} lord ${secondLord} is in ${secondPosition.rashi}; shortest Whole Sign distance is ${distance} (required ${options.requiredShortestDistance}).`;
  return {
    status: distance === options.requiredShortestDistance ? 'pass' : 'fail',
    evidence: [evidence],
  };
}

File: telugu_panchangam/personal/election_assessors/lordship.py
"""Source-neutral classical Rasi lordship and Whole Sign separation facts."""

from __future__ import annotations

from collections.abc import Mapping
from types import MappingProxyType
from typing import Any

from ...panchangam_names import RASHI_NAMES
from .contracts import PrimitiveOutcome
from .event_admission import planet_positions

LORD_POSITION_FACTS_UNAVAILABLE = (
    'Complete canonical nine-graha position facts are unavailable or conflicting.'
)

CLASSICAL_RASI_LORDS: Mapping[str, str] = MappingProxyType({
    'Mesha': 'Kuja',
    'Vrishabha': 'Shukra',
    'Mithuna': 'Budha',
    'Karka': 'Chandra',
    'Simha': 'Surya',
    'Kanya': 'Budha',
    'Tula': 'Shukra',
    'Vrischika': 'Kuja',
    'Dhanu': 'Guru',
    'Makara': 'Shani',
    'Kumbha': 'Shani',
    'Meena': 'Guru',
})


def derive_lagna_sixth_lords(
    lagna_rashi: object,
) -> tuple[str, str, str] | None:
    """Return Lagna lord, sixth Rasi, and sixth lord for a valid Lagna."""
    if lagna_rashi not in RASHI_NAMES:
        return None
    lagna_index = RASHI_NAMES.index(lagna_rashi)
    sixth_rashi = RASHI_NAMES[(lagna_index + 5) % 12]
    return (
        CLASSICAL_RASI_LORDS[lagna_rashi],
        sixth_rashi,
        CLASSICAL_RASI_LORDS[sixth_rashi],
    )


def whole_sign_shortest_distance(left_rashi: object, right_rashi: object) -> int | None:
    """Return the undirected shortest distance on the twelve-Rasi circle."""
    if left_rashi not in RASHI_NAMES or right_rashi not in RASHI_NAMES:
        return None
    delta = abs(RASHI_NAMES.index(left_rashi) - RASHI_NAMES.index(right_rashi))
    return min(delta, 12 - delta)


def evaluate_rasi_lord_separation(
    chart: Mapping[str, Any],
    *,
    first_rashi: str,
    second_rashi: str,
    required_shortest_distance: int,
) -> PrimitiveOutcome:
    """Compare two classical Rasi lords without applying an event effect."""
    if (
        first_rashi not in RASHI_NAMES
        or second_rashi not in RASHI_NAMES
        or type(required_shortest_distance) is not int
        or not 0 <= required_shortest_distance <= 6
    ):
        return PrimitiveOutcome(
            'unknown', ('The Rasi-lord separation configuration is incomplete.',)
        )
    if not isinstance(chart, Mapping):
        return PrimitiveOutcome('unknown', (LORD_POSITION_FACTS_UNAVAILABLE,))
    positions = planet_positions(chart)
    if positions is None:
        return PrimitiveOutcome('unknown', (LORD_POSITION_FACTS_UNAVAILABLE,))

    first_lord = CLASSICAL_RASI_LORDS[first_rashi]
    second_lord = CLASSICAL_RASI_LORDS[second_rashi]
    first_position = positions[first_lord]
    second_position = positions[second_lord]
    distance = whole_sign_shortest_distance(
        first_position.rashi, second_position.rashi
    )
    if distance is None:  # Defensive; admitted position facts guarantee Rasis.
        return PrimitiveOutcome('unknown', (LORD_POSITION_FACTS_UNAVAILABLE,))

    if first_lord == second_lord:
        evidence = (
            f'{first_rashi} and {second_rashi} share lord {first_lord} in '
            + f'{first_position.rashi}; shortest Whole Sign distance is 0 '
            + f'(required {required_shortest_distance}).'
        )
    else:
        evidence = (
            f'{first_rashi} lord {first_lord} is in {first_position.rashi}; '
            + f'{second_rashi} lord {second_lord} is in {second_position.rashi}; '
            + f'shortest Whole Sign distance is {distance} '
            + f'(required {required_shortest_distance}).'
        )
    return PrimitiveOutcome(
        'pass' if distance == required_shortest_distance else 'fail',
        (evidence,),
    )
